system/lib/wasmfs/syscalls.cpp

```diff
@@ -719,4 +719,133 @@ long __syscall_getdents64(long fd, long dirp, long count) {
+
+  if (oldPathParts.empty()) {
+    return -EINVAL;
+  }
+
+  // In Linux, renaming the root directory returns EBUSY.
+  // TODO: Fix this when path parsing is refactored.
+  if (oldPathParts.size() == 1 && oldPathParts[0] == "/") {
+    return -EBUSY;
+  }
+
+  auto oldBase = oldPathParts.back();
+
+  long err;
+  auto oldParentDir = getDir(oldPathParts.begin(), oldPathParts.end() - 1, err);
+
+  if (!oldParentDir) {
+    return err;
+  }
+
+  // For this operation to be atomic we must lock the source parent directory.
+  auto lockedOldParentDir = oldParentDir->locked();
+
+  auto oldPath = lockedOldParentDir.getEntry(oldBase);
+
+  if (!oldPath) {
+    return -ENOENT;
+  }
+
+  // Obtain the destination parent directory to see if it exists.
+  auto newPathParts = splitPath((char*)new_path);
+
+  if (newPathParts.empty()) {
+    return -EINVAL;
+  }
+
+  // In Linux, renaming a directory with the root directory returns ENOTEMPTY.
+  // TODO: Fix this when path parsing is refactored.
+  if (newPathParts.size() == 1 && newPathParts[0] == "/") {
+    return -ENOTEMPTY;
+  }
+
+  auto newBase = newPathParts.back();
+
+  auto newParentDir =
+    getDir(newPathParts.begin(), newPathParts.end() - 1, err, oldPath);
+
+  // If the destination parent directory doesn't exist, the source file cannot
+  // be moved.
+  if (!newParentDir) {
+    return err;
+  }
+
+  // Edge case: a/b -> a/c share the same parent. Trylocking the same parent
+  // with a recursive mutex should work, so we don't need to treat that case
+  // specially here.
+  auto maybeLockedNewParentDir = newParentDir->maybeLocked();
+  if (!maybeLockedNewParentDir) {
+    return -EBUSY;
+  }
+
+  auto lockedNewParentDir = std::move(*maybeLockedNewParentDir);
+  auto newPath = lockedNewParentDir.getEntry(newBase);
+
+  // If old_path and new_path are the same, do nothing.
+  if (newPath == oldPath) {
+    return 0;
+  }
+
+  // Cannot move to a destination parent directory wihout write permissions.
+  if (!(lockedNewParentDir.mode() & WASMFS_PERM_WRITE)) {
+    return -EACCES;
+  }
+
+  auto oldFile = oldPath->dynCast<DataFile>();
+  // The new path file must be deleted if it exists.
+  if (oldFile && newPath) {
+    // Cannot overwrite a file with a directory.
+    if (newPath->is<Directory>()) {
+      return -EISDIR;
+    }
+    lockedNewParentDir.unlinkEntry(newBase);
+  }
+
+  auto oldDirectory = oldPath->dynCast<Directory>();
+  // The new path directory must be deleted if it exists.
+  if (oldDirectory && newPath) {
+    auto newPathDirectory = newPath->dynCast<Directory>();
+
+    // Cannot overwrite a directory with a file.
+    if (!newPathDirectory) {
+      return -ENOTDIR;
+    }
+
+    // This should also cover the case in where
+    // the destination is an ancestor of the source:
+    // rename("dir/subdir", "dir");
+    if (newPathDirectory->locked().getNumEntries() > 0) {
+      return -ENOTEMPTY;
+    }
+
+    lockedNewParentDir.unlinkEntry(newBase);
+  }
```

Comment:
There is still some unnecessary duplication here: `newPath` is checked twice and unlinked twice. This structure could fix it:

```
if (newPath) {
  if (auto oldFile = oldPath->dynCast<DataFile>()) {
    // checks...
  } else if (auto oldDir = oldPath->dynCast<Directory>()) {
    // checks...
  } else {
    assert(false && "unhandled file kind");
  }
  lockedNewParentDir.unlinkEntry(newBase);
}
```

Reply:
Hmm I realize that at the end we don't even really need to casted file objects - `lockedNewParentDir.setEntry(newBase, oldPath);`

We can probably just drop this `auto oldDir = oldPath->dynCast<Directory>()`, etc.